### ISSUE TITLE
Incompatible Shapes during Validation for TensorFlow's seq2seq module

### ISSUE THREAD

**Issue by RylanSchaeffer** [tensorflow/tensorflow#12733]
I'm using TensorFlow's seq2seq module. During validation, my decoder will occasionally produce output sequences with different lengths than the target sequences; this causes the following error when calculating the loss using `tf.nn.sigmoid_cross_entropy_with_logits` (batch major, not time major):

`InvalidArgumentError (see above for traceback): Incompatible shapes: [128,22,4] vs. [128,26,4]`

What is the best practice for dealing with this problem?

I checked how the NMT tutorial solved the problem. As far as I can tell, they use a TrainingHelper during validation, which forces the decoder to unroll the same number of steps as the target sequence, but this seems like cheating - they're estimating how the model will perform during inference, but the decoder is receiving additional information (the target sequence length) that it won't have at inference time. I opened an [issue](https://github.com/tensorflow/nmt/issues/73) to clarify, but I haven't heard back.

I also [posted](https://stackoverflow.com/questions/45944131/incompatible-shapes-during-validation-for-tensorflows-seq2seq-module) on StackOverflow, but from earlier experience, I doubt I'll receive a response.

My problem isn't specific to platform or TensorFlow version, but here's that information regardless:

OS: macOS Sierra version 10.12.6
TensorFlow installed from source
TensorFlow version: ('v1.3.0-rc2-20-g0787eee', '1.3.0')

**Comment by jart** [comment]
Thank you for reaching out. You were correct to post on Stack Overflow earlier. But as you know this issue tracker is for bugs and feature requests, rather than an escalation path for Stack Overflow. We like to focus on bugs and feature requests because those types of issues, once resolved, will be beneficial to everyone in the community.

**Comment by RylanSchaeffer** [comment]
@jart , I can see why you think I'm escalating my problem, but there is an issue with TensorFlow's `Seq2Seq` module - during validation, no `Helper` subclass resolves the `Incompatible shapes` error.

I don't know if the existing helpers should solve this problem or if something new is needed, so this issue can be thought of as either a feature request or a bug. One solution is that `GreedyEmbeddingHelper`, `SampleEmbeddingHelper`, etc. should optionally take in the length of the target sequence if known (i.e. during validation) and dynamically ensure that the decoder's output matches the length of the target sequence.

The simplest solution would be to use `ScheduledOutputTrainingHelper` with `sampling_probability = 1`, but this seems to overloads the purpose of the `ScheduledOutputTrainingHelper` class.

My point regarding the NMT tutorial is that I believe the tutorial could have the same error if it didn't use a `TrainingHelper` during validation.

Please reopen the issue.

**Comment by RylanSchaeffer** [comment]
@ebrevdo @adarob Am I'm missing something incredibly obvious?

**Comment by adarob** [comment]
For validation, I either use a TrainingHelper or ScheduledOutputTrainingHelper with sampling_probabilty = 1 as you suggested. Otherwise you'll need to come up with another way of computing the loss since you will have incompatible dimensions between the inferred outputs and ground truth, as you are seeing.

**Comment by RylanSchaeffer** [comment]
I guess it doesn't really matter, but it feels wrong to use a TrainingHelper during validation.

**Comment by Daya-Jin** [comment]
Same problem. I want to do validation when traning, but meet the ```Incompatible shapes``` err. I don't know what shape the ```GreedyEmbeddingHelper``` is.

**Comment by Daya-Jin** [comment]
A stupid approach is set ```batch_size=1``` when validation.
